### DISCUSSION TITLE
About JsonDocumentation

### DISCUSSION THREAD

**Discussion by NerdFan** [Riverside-Software/pct#461] [Q&A] [answered]
Just wondering about JsonDocumentation that according release notes: "_will replace the other documentation task in next release_". 
What is idea behind? Where to consume outcome?

**Comment by gquerret** [comment, marked as answer]
Multiple goals:
* From a technical perspective, the existing documentation relies on the parser provided by Progress. As Riverside maintains its own parser for its [code analysis product](https://riverside-software.fr), cutting the dependency on a library we have no control over is a good move. There are also missing functionalities in the Progress that are available in our parser
* Also from a technical perspective, this task relies on the rcode to read the method signatures (thus with all symbols resolved, such as .Net assemblies). Doc is more accurate, and execution is much faster, as the parser only has to attach the right documentation to each method.
* Works on Linux and Windows
* There's currently no task to transform the Json output to HTML ; the developers of ABLDuck and HtmlDocumentation already said they would work on migrating their tasks to consume Json, but I think they ran out of time. Probably worth to ping them again :-)
* Both the SonarQube rules and the upcoming VSCode plugin will depend on the documentation output, in order to flag deprecated methods for example, or display documentation when hovering method calls. That probably won't directly use the Json output (in favor of some internal format), but the way documentation is generated will be common

So it might be a bit early if you want to switch to JsonDocumentation, as it can't generate Html ouput. But if you're ready to spend a bit of time, I think it's easy to adapt the ABLDuck task to use this output. PR are always welcome !

**Reply by spazzymoto** [reply]
![image](https://user-images.githubusercontent.com/2951012/165241953-4a858bf2-1b1c-42cc-aab2-408f05035ad6.png)

Made some decent progress on the a new ABLDuck version for the JsonDocumentation task. This was rewritten with vue as the original ABLDuck was written using an ancient version of sencha extjs and is not the easiest to work on.

**Reply by gquerret** [reply]
Looks good !
It would be good to display the datatype of the method parameters !

**Reply by spazzymoto** [reply]
I considered that with the first version of ABLDuck but with some methods the documentation becomes very long adding the datatype in the overview. I was going to copy what i originally did and put the data type in the body of the method documentation. Im open to suggestions though if you feel it will be better to have it in the title.

![image](https://user-images.githubusercontent.com/2951012/165679874-74c1dd4c-4875-4a58-bb83-a58b2535bc3f.png)

**Reply by gquerret** [reply]
Could you upload this sample documentation somewhere ? That would be helpful to have overall comments.

**Reply by DustinGrau-PSC** [reply]
Would the eventual intent of this combination of ABLDuck + JsonDocumentation be that the UI stands on it's own and you simply update a central JSON file with new documentation metadata? Would there be any timeframe or beta versions available to test this arrangement?

**Reply by spazzymoto** [reply]
Hey guys,
Sorry for the massive delay, between work and studying I haven't been able to find much time to work on this.

I spent some time this weekend to get it to a better state where I can start finishing it.

There is still a lot missing that i hope to add over the next few days.

@DustinGrau-PSC that is correct you would run JsonDocumentation to generate the data.json file and then just copy it to a location that is serving the ABLDuck webpage.

![image](https://user-images.githubusercontent.com/2951012/221863540-044c8c5b-40a3-40ec-8260-99a8bd56417e.png)

There will still be an ABLDuck task that will extract the version of ABLDuck compatible with the JSONDocument output and to allow you to config some custom code tags and such.

[https://spazzymoto.github.io/ablduck-next-example](https://spazzymoto.github.io/ablduck-next-example)

**Reply by gquerret** [reply]
Looks nice ! Please keep us updated :-)

**Reply by spazzymoto** [reply]
Updates:

Updated the file icons
![image](https://user-images.githubusercontent.com/2951012/222237320-7c448458-7529-4564-9fe4-d3d252d96e18.png)

Most of the work went into the comment tags system to deal with @author @deprecated @internal etc. Now you will be able to define your own tags  using something like the following

`const CUSTOM_TAGS = {
    'revision': {
        type: 'key_value',
        regex: new RegExp('^@revision\\s+(?<key>[\\d\\.]+)\\s+(?<value>.*)$', 's'),
        color: '#17802f',
        templates: {
            sidebar: 'Revision [key]',
            comment: '[key] -> [value].'
        }
    },
   'internal': {
        type: 'tag',
        color: '#600',
        templates: {
            menu: 'Int',
            signature: 'Internal',
            sidebar: 'Internal Only',
            comment: 'NOTE:This is an internal [parent] for use by the framework. Don\'t rely on its existence.'
        }
    },
}`

On the following comment
![image](https://user-images.githubusercontent.com/2951012/222240711-bef24510-946a-41cb-89a1-f39ea2f415a1.png)

Will display as follows
![image](https://user-images.githubusercontent.com/2951012/222238711-1ad3f7d6-4820-4267-a56f-8acbb149584d.png)
![image](https://user-images.githubusercontent.com/2951012/222241229-eaebf5d8-d13c-442a-9317-b663a28bb820.png)

I will expose this in a nice way in the ABLDuck task. The templates allow you to specify what to show in what location

- comment : this is shown above the code comments
- sidebar : this shows in the sidebar that floats to the right of the comment
- menu: shows a tag in the menu
- signature: shows a tag next to the signature

Supported tags types will be 

- key_value : @tag key value where key and value are named groups in your regex
- value : @tag value where value is a named group in your regex
- tag : doesnt support a regex, it is just the @tag


![image](https://user-images.githubusercontent.com/2951012/222239614-1f1b64df-d090-41be-88ab-0c1ec57995c3.png)
Worked on the filtering widget too, this should work now.

Take it for a test and let me know what you think. The Utilities & Builder class have members that have been filtered out so be sure to enable deprecate, internal and inherited for example.

https://spazzymoto.github.io/ablduck-next-example

**Comment by gquerret** [comment]
@spazzymoto I assume that this works with any Json documentation, right ? Could you publish the website for those two files ?
* https://dl.rssw.eu/corelib.json
* https://dl.rssw.eu/netlib.json

**Comment by spazzymoto** [comment]
Sure thing

- https://spazzymoto.github.io/corelib/
- https://spazzymoto.github.io/netlib/

Those have shown me a lot of places I need to work on thanks. I think I will make the base font size a bit smaller to give a bit more real estate.

Then for the long signatures I need to think, probably drop the modifier if its INPUT. This is why i think i removed the datatype in the previous version

![image](https://user-images.githubusercontent.com/2951012/222500587-ffe33eb5-fb48-4678-8b54-8c8011d4b754.png)

**Reply by PeterJudgeZA** [reply]
Looks good.

Just things I've noticed quickly
- the interfaces a type implements is not shown in the UI. The JSON does have a `interfaces` property containing this information
- interfaces can inherit from other interfaces; this is not shown in the UI. this information is show in the JSON as part of the `interfaces` and not the `inherits` property (maybe that's due to the fact that interfaces can inherit multiple interfaces).

**Reply by spazzymoto** [reply]
Ah thanks, well spotted.

A few updates, might have to ctrl-f5 to refresh cache.

Enormous lists of members scrolls left and right now
![image](https://user-images.githubusercontent.com/2951012/222682332-e273da35-8efe-48a0-b7d5-45da232d3e41.png)

Trimmed the signatures and removed INPUT (OUTPUT, INPUT-OUPUT etc should still show)
![image](https://user-images.githubusercontent.com/2951012/222682577-6a340f1b-7175-4895-a9a3-c66a9c425245.png)

Added the implements
![image](https://user-images.githubusercontent.com/2951012/222682922-74d1430f-79c3-4246-9ea7-d066e562c4c8.png)
![image](https://user-images.githubusercontent.com/2951012/222682979-9d24df22-7ed0-42fd-82e5-0b4585cffb9f.png)

**Reply by spazzymoto** [reply]
@gquerret is there any progress on the procedure part of the documentation task?

**Reply by spazzymoto** [reply]
Added some search functionality, once again refresh cache to get latest changes

![Screencast from 2023-03-03 12-47-33](https://user-images.githubusercontent.com/2951012/222701982-dd93c3db-533f-487e-9d74-260e0750514f.gif)

**Reply by spazzymoto** [reply]
The custom links to other documentation is done.

You can configure regexes to interrogate the datatype
![image](https://user-images.githubusercontent.com/2951012/222719896-1c32a9a6-920d-45ca-9f1f-99e768c3dbc8.png)

Then these will be linked in the documentation

![image](https://user-images.githubusercontent.com/2951012/222722047-7d17f5b6-1fc9-4abd-b83c-8f38ae6a730f.png)

**Reply by clement-brodu** [reply]
Hello, 

That sounds great!
When do you think you'll be able to publish a first version on github?
May be we could help ?

**Reply by clement-brodu** [reply]
Up, 

@spazzymoto, do you have any news about this ? 

We are looking forward to use it !

**Comment by spazzymoto** [comment]
Hey @clement-brodu,

There was/is some information missing from the JsonDocumentation task the last time I checked that I needed to finish off the UI.

I will see if I can put some time aside to look at implementing it.

**Reply by clement-brodu** [reply]
Feel free to ask if you need help to deal with those last issues 🙂

**Comment by clement-brodu** [comment]
Any Updates @spazzymoto ?

**Comment by spazzymoto** [comment]
Hi @clement-brodu,

Apologies work has been busy. I have 2 weeks of leave starting today, I will try dust off the project and get it committed to the abl duck repo so you can help ironing out the last few things that need to be done.
